src/Control/Carrier/Choose/Church.hs

```diff
@@ -89,8 +90,10 @@ instance MonadTrans ChooseC where
   {-# INLINE lift #-}
 
 instance (Algebra sig m, Effect sig) => Algebra (Choose :+: sig) (ChooseC m) where
-  alg (L (Choose k)) = ChooseC $ \ fork leaf -> fork (runChoose fork leaf (k True)) (runChoose fork leaf (k False))
-  alg (R other)      = ChooseC $ \ fork leaf -> alg (thread (pure ()) dst other) >>= runIdentity . runChoose (coerce fork) (coerce leaf) where
-    dst :: Applicative m => ChooseC Identity (ChooseC m a) -> m (ChooseC Identity a)
-    dst = runIdentity . runChoose (liftA2 (liftA2 (<|>))) (pure . runChoose (liftA2 (<|>)) (pure . pure))
+  alg (hom :: forall x . n x -> ChooseC m x) = \case
+    L (Choose k) -> ChooseC $ \ fork leaf -> fork (runChoose fork leaf (hom (k True))) (runChoose fork leaf (hom (k False)))
+    R other      -> ChooseC $ \ fork leaf -> alg id (thread (pure ()) dst other) >>= runIdentity . runChoose (coerce fork) (coerce leaf)
+    where
+    dst :: ChooseC Identity (n a) -> m (ChooseC Identity a)
+    dst = runIdentity . runChoose (liftA2 (liftA2 (<|>))) (pure . runChoose (liftA2 (<|>)) (pure . pure) . hom)
```

Comment:
We use `ScopedTypeVariables` now since we close over `hom`. We could probably have removed the signature for `dst` instead, but the signatures for these are a bit subtle since they use the same carrier type but containing `Identity`, so I wanted to leave the signatures intact.